pkg/cli/cmd_helpers.go

```diff
@@ -17,69 +17,331 @@ limitations under the License.
+) {
+	// In case we create a new project configuration we need to compute the layout
+	var layout string
+	if createConfig {
+		// We extract the plugin keys again instead of using the ones obtained when filtering subcommands
+		// as these plugins are unbundled but we want to keep bundle names in the layout.
+		resolvedPluginKeys := make([]string, 0, len(c.resolvedPlugins))
+		for _, p := range c.resolvedPlugins {
+			resolvedPluginKeys = append(resolvedPluginKeys, plugin.KeyFor(p))
+		}
+		layout = strings.Join(resolvedPluginKeys, ",")
+	}
+
+	options := initializationHooks(cmd, subcommands, c.metadata())
+
+	factory := executionHooksFactory{
+		fs:             c.fs,
+		store:          yamlstore.New(c.fs),
+		subcommands:    subcommands,
+		errorMessage:   errorMessage,
+		projectVersion: c.projectVersion,
+		layout:         layout,
+	}
+	cmd.PreRunE = factory.preRunEFunc(options, createConfig)
+	cmd.RunE = factory.runEFunc()
+	cmd.PostRunE = factory.postRunEFunc()
+}
+
+// initializationHooks executes update metadata and bind flags plugin hooks.
+func initializationHooks(
+	cmd *cobra.Command,
+	subcommands []keySubcommandTuple,
+	meta plugin.CLIMetadata,
+) *resourceOptions {
+	// Update metadata hook.
+	subcmdMeta := plugin.SubcommandMetadata{
+		Description: cmd.Long,
+		Examples:    cmd.Example,
+	}
+	for _, tuple := range subcommands {
+		if subcommand, updatesMetadata := tuple.subcommand.(plugin.UpdatesMetadata); updatesMetadata {
+			subcommand.UpdateMetadata(meta, &subcmdMeta)
+		}
+	}
+	cmd.Long = subcmdMeta.Description
+	cmd.Example = subcmdMeta.Examples
+
+	// Before binding specific plugin flags, bind common ones.
+	requiresResource := false
+	for _, tuple := range subcommands {
+		if _, requiresResource = tuple.subcommand.(plugin.RequiresResource); requiresResource {
+			break
+		}
+	}
+	var options *resourceOptions
+	if requiresResource {
+		options = bindResourceFlags(cmd.Flags())
+	}
+
+	// Bind flags hook.
+	for _, tuple := range subcommands {
+		if subcommand, hasFlags := tuple.subcommand.(plugin.HasFlags); hasFlags {
+			subcommand.BindFlags(cmd.Flags())
+		}
+	}
+
+	return options
+}
+
+type executionHooksFactory struct {
+	// fs is the filesystem abstraction to scaffold files to.
+	fs machinery.Filesystem
+	// store is the backend used to load/save the project configuration.
+	store store.Store
+	// subcommands are the tuples representing the set of subcommands provided by the resolved plugins.
+	subcommands []keySubcommandTuple
+	// errorMessage is prepended to returned errors.
+	errorMessage string
+	// projectVersion is the project version that will be used to create new project configurations.
+	// It is only used for initialization.
+	projectVersion config.Version
+	// layout is the plugin chain configured for this project.
+	layout string
+}
+
+// subcommandTuples iterates over the subcommands skipping those flagged to do so.
+func (factory *executionHooksFactory) subcommandTuples() chan keySubcommandTuple {
```

Comment:
I don't want to make this too absurd, but is there any reason we can't simplify these three methods down to

```go
func newFilter(subcmds []plugin.Subcommand) {
  keys := /* extract keys */
  return &filteredSubcommands {
    subcommands: subcmds,
    keys: keys,
    skip: make([]bool, len(subcmds)),
  }
}
type filteredSubcommands struct {
  subcommands []plugin.Subcommand
  keys []string
  skip []bool
}
func (s *filteredSubcommands) ForEach(cb func(subcmd plugin.Subcommand) error) error {
  for i, cmd := range s.subcommands {
    if s.skip[i] {
        continue
    }
    
    err := cb(cmd)
    if err != nil {
        var exitError plugin.ExitError
        if errors.As(err, &exitError) {
          s.skip[i] = true
          fmt.Printf("skipping remaining hooks for %q: %s\n", s.keys[i], exitError.Reason)
          continue
        }
        return fmt.Errorf("for %s, %w", s.keys[i], err)
    }
  }
  return nil
}
````
 
Then below, using it becomes

```go
filterer := newFilter(subcommands)

// ...

err := filterer.ForEach(func(subcmd plugin.Subcommand) error {
  if subcmd, reqCfg := subcmd.(plugin.RequiresConfig); reqCfg {
    return subcmd.InjectConfig(cfg)
  }
  return nil
})
if err != nil {
  return fmt.Errorf("unable to inject configuration: %w", err)
}

// etc
```

Reply:
Will apply this approach if I don't see any drawback and rebase

Reply:
Done